Issue #8023: Update grammar for initial record support

nmancus1 opened the pull request:
Issue #8023: Update grammar for initial record support
### TO DO:
- [x] Unit tests (including ASTRegressionTest after https://github.com/checkstyle/checkstyle/pull/8290# is merged)

- [x] Modify grammar to avoid "record" in variable declaration:
https://github.com/checkstyle/checkstyle/blob/2c33bb063dea4be60b131c8d593d2dc074e90855/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14Records.java#L103-L104


- [x] Modify grammar to handle no arg record constructor:
https://github.com/checkstyle/checkstyle/blob/2c33bb063dea4be60b131c8d593d2dc074e90855/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14Records.java#L46-L49


_Mentors, please feel free to edit this post to add more items in TODO, or reply and I will add them._

nmancus1 commented:
This post will be for AST structure regression report and check behavior regression reports when they are complete.

esilkensen commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14Records.java:
```diff
@@ -24,15 +35,15 @@ public record AnnotatedBinaryNode(@NonNull1 Node left, @NonNull1 Node right) { }
 
     //Implements interface
     public record Polar(double r, double theta) implements Coords {
-        @Override
-        public double x() {
-            return r * Math.cos(theta);
-        }
+    @Override
```
looks like this record's indentation got off (I assume unintentionally?)

nmancus1 replied:
Fixed

esilkensen commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -1966,6 +2014,17 @@ IDENT
+                && LA(1) != ')' && LA(1) != '.'
+                && LA(1) != ',' && LA(1) != '('
+                && LA(2) != ':') {
+                $setType(RECORD);
```
I wondered about an approach like this as well -- I think it may be hard to catch all the cases, though (e.g. `record[]`, `record ->`, `record<T>`, etc.), or to really decide keyword vs. identifier in the lexer (what about `class record extends Object`), and seems like it could be difficult to maintain.

It makes me wonder again about leaving `record` as an `IDENT` as far as the lexer is concerned, and just trying to deal with it in the grammar rule. I know you have already been down multiple paths with predicates -- I will take a look myself a little later tonight and see if anything strikes me...

nmancus1 replied:
[This](https://github.com/nmancus1/checkstyle/commit/7c9102b62960fa1f1b3790251aa8bc9045aeaafc) actually works pretty great for records, but messes up quite a few other things.  I got the idea from [here](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fjszheng%2Fpy3antlr4book%2Fblob%2Fmaster%2F12-id_as_keyword%2FPredKeyword.g4&sa=D&sntz=1&usg=AFQjCNFN0TC9Wyh_WM_05f3ZSfTWJu6tYQ).

esilkensen commented on src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:
```diff
@@ -3539,6 +3539,29 @@
+     * The {@code record} keyword.  This element appears
+     * as part of a record declaration.
+     **/
+    public static final int LITERAL_RECORD =
+            GeneratedJavaTokenTypes.LITERAL_record;
```
It doesn't look this is referenced in the grammar. Can it be removed?

nmancus1 commented:
I am reducing the minimum coverage for` GeneratedJavaRecognizer.java` from 0.73 to 0.72. The reason for this is because the code generated by the new grammar rules I have introduced have adequate coverage.  There are only two rules that don't have 100% coverage. These are:

**`recordComponents`**  

![image](https://user-images.githubusercontent.com/31252532/84401665-87052f80-abd1-11ea-9caa-d6bbcef9808f.png)
This amount of coverage is typical for rules that use `t:typeSpec[false] `, like `enumConstantField`.

**`recordDeclaration`**

![image](https://user-images.githubusercontent.com/31252532/84404461-d4cf6700-abd4-11ea-886c-11f369cb131b.png)

Other than the default option, the only branch not covered is the false comparison result of the ternary statment on line 818, and the same branch in `classDefinition` (which is very similar to recordDeclaration) is also not covered:
![image](https://user-images.githubusercontent.com/31252532/84404992-866e9800-abd5-11ea-91db-9ac8e43f66f9.png)


Below is the coverage for all of the rules I've added:

`![image](https://user-images.githubusercontent.com/31252532/84398854-70a9a480-abce-11ea-8a99-03655e76be80.png)
![image](https://user-images.githubusercontent.com/31252532/84398935-8cad4600-abce-11ea-99b8-98c003270cd5.png)

rnveach commented:
@nmancus1 FYI on how to hit the harder to reach exceptions: https://github.com/checkstyle/checkstyle/blob/dabe0ff1576239d5623807f787bedea1020bbb23/src/test/java/com/puppycrawl/tools/checkstyle/grammar/AstRegressionTest.java#L201

nmancus1 commented:
Execution time testing on guava with `google_checks.xml` (I've omitted the other time outputs and just kept the total):

`➜  checkstyle-tester git:(master) ✗ for run in {1..10}; do time groovy diff.groovy --localGitRepo /home/nick/IdeaProjects/checkstyle --patchBranch master --patchConfig ../../google_checks.xml --listOfProjects projects-to-test-on.properties --mode single |& grep 'Total time:'; done`
```
[INFO] Total time:  42.857 s
[INFO] Total time:  43.172 s
[INFO] Total time:  42.233 s
[INFO] Total time:  43.812 s
[INFO] Total time:  42.149 s
[INFO] Total time:  44.875 s
[INFO] Total time:  46.218 s
[INFO] Total time:  43.659 s
[INFO] Total time:  45.105 s
[INFO] Total time:  42.515 s
[INFO] Total time:  43.349 s
[INFO] Total time:  45.786 s
[INFO] Total time:  45.326 s
[INFO] Total time:  43.102 s
[INFO] Total time:  43.368 s
[INFO] Total time:  43.545 s
[INFO] Total time:  44.725 s
[INFO] Total time:  44.314 s
[INFO] Total time:  46.607 s
[INFO] Total time:  43.914 s
```
`➜  checkstyle-tester git:(master) ✗ for run in {1..10}; do time groovy diff.groovy --localGitRepo /home/nick/IdeaProjects/checkstyle --patchBranch initial-record-grammar --patchConfig ../../google_checks.xml --listOfProjects projects-to-test-on.properties --mode single |& grep 'Total time:'; done`

```
[INFO] Total time:  44.310 s
[INFO] Total time:  43.649 s
[INFO] Total time:  42.931 s
[INFO] Total time:  43.118 s
[INFO] Total time:  44.582 s
[INFO] Total time:  44.762 s
[INFO] Total time:  43.479 s
[INFO] Total time:  43.130 s
[INFO] Total time:  43.646 s
[INFO] Total time:  42.599 s
[INFO] Total time:  42.074 s
[INFO] Total time:  42.578 s
[INFO] Total time:  42.702 s
[INFO] Total time:  42.881 s
[INFO] Total time:  42.860 s
[INFO] Total time:  44.799 s
[INFO] Total time:  45.955 s
[INFO] Total time:  43.048 s
[INFO] Total time:  43.801 s
[INFO] Total time:  43.279 s
```
Which gives an average time of 44.031s for `master`, and 43.509s for `initial-record-grammar`.

nmancus1 commented:
> @nmancus1 FYI on how to hit the harder to reach exceptions:
@rnveach 
Unless I'm missing something, these tests are for the lexer only.

esilkensen commented on pom.xml:
```diff
@@ -906,7 +906,7 @@
                       <counter>BRANCH</counter>
                       <value>COVEREDRATIO</value>
-                      <minimum>0.73</minimum>
+                      <minimum>0.72</minimum>
```
nit: looks like the commit message references the wrong file (Lexer instead of Recognizer)

nmancus1 replied:
Good catch. Fixed.

nmancus1 commented:
With these grammar changes, there has been some additional nondeterminism detected:
![image](https://user-images.githubusercontent.com/31252532/84518151-48897680-ac9e-11ea-915f-242c013d6529.png)
Sans line numbers:
![image](https://user-images.githubusercontent.com/31252532/84519504-31e41f00-aca0-11ea-9294-589b480abbc1.png)

Aside from where nondeterminism existed previously, these are the rules in which it has been inroduced: 
```
java.g:443: warning:nondeterminism upon
java.g:443:     k==1:DOT
java.g:443:     k==2:"record"
java.g:443:     between alt 1 and exit branch of block

java.g:951: warning:nondeterminism between alts 1 and 2 of block upon
java.g:951:     k==1:AT
java.g:951:     k==2:RBRACK,IDENT,"record"

java.g:761:9: warning:nondeterminism between alts 1 and 2 of block upon
java.g:761:9:     k==1:AT,"record"
java.g:761:9:     k==2:IDENT,"record"

java.g:1045:31: warning:nondeterminism between alts 1 and 2 of block upon
java.g:1045:31:     k==1:IDENT,"record"
java.g:1045:31:     k==2:LBRACK,DOT,AT
```
These lines correspond to the `identifierStar`, `declaratorBrackets`, `enumConstantField`, and `parameterDeclaration` rules, respectively.

esilkensen commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -254,6 +257,7 @@ typeDefinitionInternal[AST modifiers]
     | interfaceDefinition[#modifiers]
     | enumDefinition[#modifiers]
     | annotationDefinition[#modifiers]
+    | recordDeclaration[#modifiers]
```
What do you think about naming this `recordDefinition` to fit in with `enumDefinition`, `annotationDefinition`, etc.? (If you think that makes sense, no need to rerun reports just for that renaming)

pbludov replied:
+1 for `recordDefinition`. Records are basically classes.

nmancus1 replied:
I was thinking the same. Done.

esilkensen commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,9 +547,60 @@ annotationExpression
+    :    annotations modifiers id
+            constructorBody
+    ;
+//
```
intentional?

nmancus1 replied:
It was, to sort of group all of the rules related to #8267 together.  I've removed it.

esilkensen commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,9 +547,60 @@ annotationExpression
+                              modifiers, r, id, rc);}
+    ;
+
+// Build no AST for below rules until full records support
```
Could potentially link to #8267

nmancus1 replied:
Done.

pbludov commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -1693,7 +1757,6 @@ options {
     {
         mTreatEnumAsKeyword = aTreatAsKeyword;
     }
-
```
This blank line has a pair on line 1668 (1732). Please remove both of them or none of them.

nmancus1 replied:
Fixed.

pbludov commented:
@nmancus1 please rebase to fix the Wercker build.

rnveach commented:
Can someone post the JLS link for records and how they specify the language?

Also I notice this isn't full record support. Is there another issue defined that will house the full record support?

esilkensen commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,9 +547,61 @@ annotationExpression
+    ;
+
+// Build no AST for other record rules until
+// https://github.com/checkstyle/checkstyle/issues/8267
```
@rnveach 
>Also I notice this isn't full record support. Is there another issue defined that will house the full record support?

Here is a link to the issue for full record support

nmancus1 commented:
>Can someone post the JLS link for records and how they specify the language?

See https://github.com/checkstyle/checkstyle/issues/8023#issuecomment-636608784

rnveach commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,6 +547,58 @@ annotationExpression
+        ic:implementsClause
+        rb:recordBodyDeclaration
+        {#recordDefinition = #(#[RECORD_DEF, "RECORD_DEF"],
+                              modifiers, r, id, rc);}
```
we not adding tp, ic rb because it is incomplete? I noticed the grammar print had nothing.

nmancus1 replied:
>we not adding tp, ic rb because it is incomplete?

That is correct, but I still needed to include them here to correctly parse everything.

rnveach replied:
Ok, it may just be me but it sounds like it would be beneficial to add a comment saying updates will come in 8267 to ensure we don't miss anything.

nmancus1 replied:
I will make it more clear.

rnveach commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,6 +547,58 @@ annotationExpression
+    :   LCURLY
+        (   (compactConstructorDeclaration)=> compactConstructorDeclaration
+        |   field
+        |   SEMI
```
http://cr.openjdk.java.net/~gbierman/jep359/jep359-20191031/specs/records-jls.html#jls-8.10.2

esilkensen replied:
I think we are covered with `field` and `SEMI` (I like this comment 😄) https://github.com/checkstyle/checkstyle/blob/f1ee17d45801f57c4a20a11d1311116c0e1d946c/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g#L763-L768

This PR is parsing the [InputJava14Records](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14Records.java) file which has records with fields, methods, etc. in their bodies.

nmancus1 replied:
>Is there no ClassBodyDeclaration because of limited record support?

That was my original intention, but I don't see why we shouldn't go ahead and add it now.

rnveach replied:
I can't provide recommendation on best way to go. If you believe this is parsing correctly or will be changed in full support, then you can mark this as resolved. I just noticed the disparity and wanted to make sure I was missing something or it was going to be resolved.

nmancus1 replied:
This parses everything I've thrown at it, and all the records files in JDK14 repo.  The only changes I plan to make are to build the AST for it in #8267.  I'm going to mark it as resolved.

nmancus1 commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,6 +547,58 @@ annotationExpression
 
+
+recordDefinition![AST modifiers]
+    :   r:"record" id:id
```
Yes, it will work.

nmancus1 replied:
Fixed.

rnveach replied:
Please make the change if you are ok with it. To me it seems odd we create the literal but don't use it. I noticed this other times searching around so its not a big deal.

nmancus1 commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -116,6 +116,9 @@ tokens {
     FLOAT_LITERAL; DOUBLE_LITERAL; HEX_FLOAT_LITERAL; HEX_DOUBLE_LITERAL;
     SIGNED_INTEGER; BINARY_EXPONENT;
+
+    //Start Java14 syntax here
```
Fixed.

rnveach commented:
I think this is ok to merge when CI passes. @romani

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:
```diff
@@ -3540,6 +3540,34 @@
+     *
+     * <p>For example:</p>
+     * <pre>
+     * public record myRecord (String myString, String myOtherString) {}
```
I do not see `myString` in AST example.
Please modify code example to be minimalistic, AST should be same by structure as our CLI is printing it.

nmancus1 replied:
Fixed, I also updated the format of the AST to match the format we discussed in other PR.

romani commented on src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:
```diff
@@ -543,6 +546,59 @@ annotationExpression
+// until https://github.com/checkstyle/checkstyle/issues/8267
+recordDefinition![AST modifiers]
+    :   r:LITERAL_record id:id
+        (tp:typeParameters)?        //until #8267
```
`// until ...` space after `//`

nmancus1 replied:
Fixed.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:
```diff
@@ -3540,6 +3540,32 @@
+            GeneratedJavaTokenTypes.LITERAL_record;
+
+    /**
+     * A record declaration.
```
as we skip ``--OBJBLOCK -> OBJBLOCK` in AST, lets update `A record declaration` to `A record declaration, this implementation is just to avoid parse errors, full support will be at #8267`

nmancus1 replied:
Fixed and rebased.

romani commented:
there are conflicts, please rebase and prepare for merge.
